lib/utils.js

```diff
@@ -144,38 +144,48 @@ exports.encode = function (str) {
     return out;
 };
 
-exports.compact = function (obj, references) {
-    if (typeof obj !== 'object' || obj === null) {
-        return obj;
+exports.compact = function (value) {
+    var queue = [{ obj: { o: value }, prop: 'o' }];
+    var refs = [];
+
+    for (var i = 0; i < queue.length; ++i) {
+        var item = queue[i];
+        var obj = item.obj[item.prop];
+
+        var keys = Object.keys(obj);
+        for (var j = 0; j < keys.length; ++j) {
+            var key = keys[j];
+            var val = obj[key];
+            if (typeof val === 'object' && val !== null && refs.indexOf(val) === -1) {
+                queue.push({ obj: obj, prop: key });
+                refs.push(val);
+            }
+        }
     }
 
-    var refs = references || [];
-    var lookup = refs.indexOf(obj);
-    if (lookup !== -1) {
-        return refs[lookup];
-    }
+    return exports.compactQueue(queue);
+};
+
+exports.compactQueue = function (queue) {
```

Comment:
does this need to be publicly exported? It'd be nice not to increase the public API.

Reply:
It doesn't, but didn't see the `utils` object being exported out through the `index` entry point. I wasn't sure if adding to `exports` was just the style or not (most of it was changed to the Hapi project style when it lived there for a bit and I just wasn't super familiar).

TL;DR thought that it would fit in that way, but I'll in-export the function 👍

Reply:
It's certainly consistent :-) but I'm in a slow yet determined march to move away from the hapi style, and minimize the publicly reachable API of `qs`, so the change is appreciated!

Reply:
haha, making the change is no problem at all :) I try to copy the existing code style as much as possible to blend it in, so wasn't sure, hehe